File: src/visual_generator.py
"""
Visual Generator - Creates procedural visual elements
"""

import pygame
import math
import random
from typing import List, Tuple
# Commenting out PIL and numpy for now to avoid dependencies
# from PIL import Image, ImageDraw, ImageFilter
# import numpy as np


class VisualGenerator:
    """Generates procedural visual elements for the game"""
    
    def __init__(self):
        self.noise_cache = {}
        
    def generate_background_texture(self, width: int, height: int, 
                                  seed: int = 0) -> pygame.Surface:
        """Generate a subtle background texture"""
        # Create pygame surface directly
        surface = pygame.Surface((width, height))
        surface.fill((10, 10, 15))
        
        random.seed(seed)
        
        # Add noise points
        for _ in range(1000):
            x = random.randint(0, width - 1)
            y = random.randint(0, height - 1)
            brightness = random.randint(0, 30)
            color = (brightness, brightness, brightness + 10)
            surface.set_at((x, y), color)
            
        # Add subtle gradient
        for y in range(0, height, 5):
            gradient_strength = int(20 * (y / height))
            for x in range(0, width, 10):
                color = (
                    10 + gradient_strength,
                    10 + gradient_strength,
                    15 + gradient_strength
                )
                if x < width and y < height:
                    surface.set_at((x, y), color)
        
        random.seed()
        return surface
        
    def generate_connection_line(self, start: Tuple[int, int], end: Tuple[int, int],
                               strength: float, color: Tuple[int, int, int]) -> pygame.Surface:
        """Generate a stylized connection line between two points"""
        # Calculate line dimensions
        dx = end[0] - start[0]
        dy = end[1] - start[1]
        length = math.sqrt(dx * dx + dy * dy)
        
        if length < 1:
            return pygame.Surface((1, 1))
        
        # Create surface for the line
        line_width = max(1, int(strength * 5))
        surface = pygame.Surface((int(length) + line_width * 2, line_width * 2 + 10), 
                                pygame.SRCALPHA)
        
        # Draw the line with gradient
        center_y = surface.get_height() // 2
        
        for x in range(int(length)):
            # Calculate line intensity (stronger in middle)
            intensity = 1 - abs(x - length/2) / (length/2)
            intensity = max(0.2, intensity)
            
            # Draw vertical line segment
            for y in range(line_width):
                alpha = int(255 * strength * intensity)
                pixel_color = (*color, alpha)
                
                # Add slight wave effect
                wave_y = center_y + int(math.sin(x * 0.1) * 2)
                if 0 <= wave_y + y < surface.get_height():
                    surface.set_at((x, wave_y + y - line_width//2), pixel_color)
        
        return surface
        
    def generate_memory_aura(self, size: int, color: Tuple[int, int, int], 
                           intensity: float) -> pygame.Surface:
        """Generate a glowing aura around memory fragments"""
        surface = pygame.Surface((size * 2, size * 2), pygame.SRCALPHA)
        center = size
        
        # Generate multiple layers of glow
        for layer in range(5):
            radius = size - layer * (size // 6)
            if radius <= 0:
                break
                
            # Calculate alpha for this layer
            alpha = int(intensity * 60 / (layer + 1))
            
            # Create glow color
            glow_color = (
                min(255, color[0] + 50),
                min(255, color[1] + 50),
                min(255, color[2] + 50),
                alpha
            )
            
            # Draw the glow ring
            pygame.draw.circle(surface, glow_color, (center, center), radius, 2)
            
        return surface
        
    def generate_healing_effect(self, x: int, y: int, progress: float,
                              color: Tuple[int, int, int]) -> pygame.Surface:
        """Generate visual effect for healing process"""
        size = 100
        surface = pygame.Surface((size, size), pygame.SRCALPHA)
        center = size // 2
        
        # Draw expanding rings
        for i in range(3):
            ring_progress = (progress + i * 0.3) % 1.0
            radius = int(ring_progress * center)
            alpha = int(255 * (1 - ring_progress))
            
            if radius > 0:
                ring_color = (*color, alpha)
                pygame.draw.circle(surface, ring_color, (center, center), radius, 2)
        
        # Draw healing particles
        for i in range(8):
            angle = (math.pi * 2 * i) / 8 + progress * math.pi
            particle_x = center + int(math.cos(angle) * 30)
            particle_y = center + int(math.sin(angle) * 30)
            
            particle_color = (*color, int(255 * (1 - progress)))
            pygame.draw.circle(surface, particle_color, (particle_x, particle_y), 3)
            
        return surface
        
    def generate_level_transition(self, width: int, height: int, 
                                progress: float) -> pygame.Surface:
        """Generate transition effect between levels"""
        surface = pygame.Surface((width, height), pygame.SRCALPHA)
        
        # Create wave effect
        for x in range(width):
            wave_height = int(math.sin(x * 0.02 + progress * math.pi * 2) * 50)
            start_y = height // 2 + wave_height
            
            for y in range(start_y, height):
                distance = y - start_y
                alpha = max(0, int(255 * (1 - distance / 100) * progress))
                
                if alpha > 0:
                    color = (100, 150, 255, alpha)
                    surface.set_at((x, y), color)
        
        return surface
        
    def generate_noise_texture(self, width: int, height: int, 
                             scale: float = 0.1) -> pygame.Surface:
        """Generate Perlin-like noise texture"""
        surface = pygame.Surface((width, height))
        
        for x in range(width):
            for y in range(height):
                # Simple noise function
                noise_val = math.sin(x * scale) * math.sin(y * scale)
                noise_val += math.sin(x * scale * 2) * math.sin(y * scale * 2) * 0.5
                noise_val += math.sin(x * scale * 4) * math.sin(y * scale * 4) * 0.25
                
                # Normalize to 0-1
                noise_val = (noise_val + 1.875) / 3.75
                
                # Convert to grayscale
                gray = int(noise_val * 255)
                surface.set_at((x, y), (gray, gray, gray))
        
        return surface

File: renderer/js/visual-generator.js
// Visual Generator for Memory Drift

class VisualGenerator {
    constructor() {
        this.noiseCache = new Map();
    }
    
    generateBackgroundTexture(canvas, seed = 0) {
        const ctx = canvas.getContext('2d');
        const width = canvas.width;
        const height = canvas.height;
        
        // Clear canvas
        ctx.fillStyle = '#0a0a0f';
        ctx.fillRect(0, 0, width, height);
        
        // Add noise points
        const imageData = ctx.getImageData(0, 0, width, height);
        const data = imageData.data;
        
        for (let i = 0; i < 1000; i++) {
            const x = Math.floor(Utils.seededRandom(seed + i) * width);
            const y = Math.floor(Utils.seededRandom(seed + i + 1000) * height);
            const brightness = Math.floor(Utils.seededRandom(seed + i + 2000) * 30);
            
            const index = (y * width + x) * 4;
            if (index >= 0 && index < data.length - 3) {
                data[index] = brightness;     // R
                data[index + 1] = brightness; // G
                data[index + 2] = brightness + 10; // B
                data[index + 3] = 255; // A
            }
        }
        
        // Add subtle gradient
        for (let y = 0; y < height; y += 5) {
            const gradientStrength = Math.floor(20 * (y / height));
            for (let x = 0; x < width; x += 10) {
                const index = (y * width + x) * 4;
                if (index >= 0 && index < data.length - 3) {
                    data[index] = 10 + gradientStrength;     // R
                    data[index + 1] = 10 + gradientStrength; // G
                    data[index + 2] = 15 + gradientStrength; // B
                    data[index + 3] = 255; // A
                }
            }
        }
        
        ctx.putImageData(imageData, 0, 0);
    }
    
    drawConnectionLine(ctx, start, end, strength, color) {
        const dx = end.x - start.x;
        const dy = end.y - start.y;
        const length = Math.sqrt(dx * dx + dy * dy);
        
        if (length < 1) return;
        
        const lineWidth = Math.max(1, strength * 5);
        
        ctx.save();
        ctx.strokeStyle = Utils.colorToString({
            ...color,
            a: strength
        });
        ctx.lineWidth = lineWidth;
        
        // Add slight wave effect
        const segments = Math.floor(length / 10);
        if (segments > 1) {
            ctx.beginPath();
            ctx.moveTo(start.x, start.y);
            
            for (let i = 1; i < segments; i++) {
                const t = i / segments;
                const x = Utils.lerp(start.x, end.x, t);
                const y = Utils.lerp(start.y, end.y, t);
                
                // Add wave
                const waveOffset = Math.sin(t * Math.PI * 2) * 2 * strength;
                const perpX = -dy / length;
                const perpY = dx / length;
                
                ctx.lineTo(x + perpX * waveOffset, y + perpY * waveOffset);
            }
            
            ctx.lineTo(end.x, end.y);
            ctx.stroke();
        } else {
            ctx.beginPath();
            ctx.moveTo(start.x, start.y);
            ctx.lineTo(end.x, end.y);
            ctx.stroke();
        }
        
        ctx.restore();
    }
    
    drawMemoryAura(ctx, x, y, size, color, intensity) {
        ctx.save();
        
        // Draw multiple layers of glow
        for (let layer = 0; layer < 5; layer++) {
            const radius = size - layer * (size / 6);
            if (radius <= 0) break;
            
            const alpha = (intensity * 0.6) / (layer + 1);
            
            ctx.strokeStyle = Utils.colorToString({
                r: Math.min(255, color.r + 50),
                g: Math.min(255, color.g + 50),
                b: Math.min(255, color.b + 50),
                a: alpha
            });
            ctx.lineWidth = 2;
            
            ctx.beginPath();
            ctx.arc(x, y, radius, 0, Math.PI * 2);
            ctx.stroke();
        }
        
        ctx.restore();
    }
    
    drawHealingEffect(ctx, x, y, progress, color) {
        ctx.save();
        
        const size = 50;
        
        // Draw expanding rings
        for (let i = 0; i < 3; i++) {
            const ringProgress = (progress + i * 0.3) % 1.0;
            const radius = ringProgress * size;
            const alpha = 1 - ringProgress;
            
            if (radius > 0) {
                ctx.strokeStyle = Utils.colorToString({
                    ...color,
                    a: alpha
                });
                ctx.lineWidth = 2;
                
                ctx.beginPath();
                ctx.arc(x, y, radius, 0, Math.PI * 2);
                ctx.stroke();
            }
        }
        
        // Draw healing particles
        for (let i = 0; i < 8; i++) {
            const angle = (Math.PI * 2 * i) / 8 + progress * Math.PI;
            const particleX = x + Math.cos(angle) * 30;
            const particleY = y + Math.sin(angle) * 30;
            
            ctx.fillStyle = Utils.colorToString({
                ...color,
                a: 1 - progress
            });
            
            ctx.beginPath();
            ctx.arc(particleX, particleY, 3, 0, Math.PI * 2);
            ctx.fill();
        }
        
        ctx.restore();
    }
    
    drawLevelTransition(ctx, width, height, progress) {
        ctx.save();
        
        // Create wave effect
        const gradient = ctx.createLinearGradient(0, 0, 0, height);
        gradient.addColorStop(0, `rgba(100, 150, 255, 0)`);
        gradient.addColorStop(progress, `rgba(100, 150, 255, ${progress * 0.8})`);
        gradient.addColorStop(1, `rgba(100, 150, 255, ${progress})`);
        
        ctx.fillStyle = gradient;
        
        // Draw wave
        ctx.beginPath();
        ctx.moveTo(0, height);
        
        for (let x = 0; x <= width; x += 10) {
            const waveHeight = Math.sin(x * 0.02 + progress * Math.PI * 2) * 50;
            const y = height / 2 + waveHeight;
            ctx.lineTo(x, y);
        }
        
        ctx.lineTo(width, height);
        ctx.lineTo(0, height);
        ctx.closePath();
        ctx.fill();
        
        ctx.restore();
    }
    
    generateNoiseTexture(width, height, scale = 0.1) {
        const canvas = document.createElement('canvas');
        canvas.width = width;
        canvas.height = height;
        const ctx = canvas.getContext('2d');
        
        const imageData = ctx.getImageData(0, 0, width, height);
        const data = imageData.data;
        
        for (let x = 0; x < width; x++) {
            for (let y = 0; y < height; y++) {
                // Simple noise function
                let noiseVal = Math.sin(x * scale) * Math.sin(y * scale);
                noiseVal += Math.sin(x * scale * 2) * Math.sin(y * scale * 2) * 0.5;
                noiseVal += Math.sin(x * scale * 4) * Math.sin(y * scale * 4) * 0.25;
                
                // Normalize to 0-1
                noiseVal = (noiseVal + 1.875) / 3.75;
                
                // Convert to grayscale
                const gray = Math.floor(noiseVal * 255);
                const index = (y * width + x) * 4;
                
                data[index] = gray;     // R
                data[index + 1] = gray; // G
                data[index + 2] = gray; // B
                data[index + 3] = 255;  // A
            }
        }
        
        ctx.putImageData(imageData, 0, 0);
        return canvas;
    }
    
    // Pattern generation utilities
    generateCirclePattern(count, centerX, centerY, radius) {
        const positions = [];
        for (let i = 0; i < count; i++) {
            const angle = (2 * Math.PI * i) / count;
            const x = centerX + radius * Math.cos(angle);
            const y = centerY + radius * Math.sin(angle);
            positions.push({ x, y });
        }
        return positions;
    }
    
    generateSpiralPattern(count, centerX, centerY) {
        const positions = [];
        for (let i = 0; i < count; i++) {
            const angle = (2 * Math.PI * i) / 5;
            const radius = 50 + i * 25;
            const x = centerX + radius * Math.cos(angle);
            const y = centerY + radius * Math.sin(angle);
            positions.push({ x, y });
        }
        return positions;
    }
    
    generateGridPattern(count, width, height) {
        const positions = [];
        const gridSize = Math.ceil(Math.sqrt(count));
        const spacingX = width / (gridSize + 1);
        const spacingY = height / (gridSize + 1);
        
        for (let i = 0; i < count; i++) {
            const row = Math.floor(i / gridSize);
            const col = i % gridSize;
            const x = (col + 1) * spacingX;
            const y = (row + 1) * spacingY;
            positions.push({ x, y });
        }
        return positions;
    }
    
    generateConstellationPattern(count, width, height) {
        const positions = [];
        const margin = 100;
        
        for (let i = 0; i < count; i++) {
            let attempts = 0;
            let x, y;
            
            do {
                x = Utils.random(margin, width - margin);
                y = Utils.random(margin, height - margin);
                attempts++;
            } while (attempts < 50 && positions.some(pos => 
                Utils.distance(pos, { x, y }) < 80
            ));
            
            positions.push({ x, y });
        }
        return positions;
    }
    
    generateMandalaPattern(count, centerX, centerY) {
        const positions = [];
        const rings = 3;
        const itemsPerRing = Math.ceil(count / rings);
        
        for (let i = 0; i < count; i++) {
            const ring = Math.floor(i / itemsPerRing);
            const radius = 80 + ring * 80;
            const angleStep = (2 * Math.PI) / itemsPerRing;
            const angle = (i % itemsPerRing) * angleStep;
            
            const x = centerX + radius * Math.cos(angle);
            const y = centerY + radius * Math.sin(angle);
            positions.push({ x, y });
        }
        return positions;
    }
}